src/main/java/sh/platform/config/reader/CredentialFormatter.java

```diff
@@ -3,5 +3,14 @@
 import java.util.Map;
 import java.util.function.Function;
 
+/**
+ * In some cases the library being used to connect to a service wants its credentials formatted in a specific way;
+ * it could be a DSN string of some sort or it needs certain values concatenated to the database name, etc.
+ * For those cases you can use "Credential Formatters".
+ * A Credential Formatter is a functional interface that takes a credentials array and returns any type, since the
+ * library may want different types.
+ *
+ * @param <T> the produced type
+ */
 public interface CredentialFormatter<T> extends Function<Map<String, Object>, T> {
```

Comment:
Java can define function-objects now?  Objects that are themselves callable?  Very cool!

One thing I'm not sure on here: At least in other languages in concept the formatter may return a string OR a dict, because different libraries want different things.  Does the generic here allow for that?

Reply:
Yes, interface with one method is not an interface function that allows lambda and when combining with generics we can return anything :)